File: 04-stateful-stream-processing/java-scala/e2/src/main/java/com/becomedataengineer/StatefulFlinkConsumer.java
package com.becomedataengineer;

import com.fasterxml.jackson.core.JsonProcessingException;
import org.apache.flink.api.common.eventtime.WatermarkStrategy;
import org.apache.flink.api.common.serialization.SimpleStringSchema;
import org.apache.flink.api.java.utils.ParameterTool;
import org.apache.flink.configuration.Configuration;
import org.apache.flink.connector.base.DeliveryGuarantee;
import org.apache.flink.connector.kafka.sink.KafkaRecordSerializationSchema;
import org.apache.flink.connector.kafka.sink.KafkaSink;
import org.apache.flink.connector.kafka.source.KafkaSource;
import org.apache.flink.connector.kafka.source.enumerator.initializer.OffsetsInitializer;
import org.apache.flink.streaming.api.datastream.SingleOutputStreamOperator;
import org.apache.flink.streaming.api.datastream.WindowedStream;
import org.apache.flink.streaming.api.environment.CheckpointConfig;
import org.apache.flink.streaming.api.environment.StreamExecutionEnvironment;
import org.apache.flink.streaming.api.functions.windowing.ProcessWindowFunction;
import org.apache.flink.streaming.api.windowing.assigners.TumblingEventTimeWindows;
import org.apache.flink.streaming.api.windowing.time.Time;
import org.apache.flink.streaming.api.windowing.triggers.CountTrigger;
import org.apache.flink.streaming.api.windowing.triggers.EventTimeTrigger;
import org.apache.flink.streaming.api.windowing.windows.TimeWindow;
import org.apache.flink.streaming.api.windowing.windows.Window;
import org.apache.flink.util.Collector;
import org.apache.flink.util.IterableUtils;

import java.time.Instant;
import java.time.LocalDateTime;
import java.time.ZoneId;
import java.util.HashMap;
import java.util.Map;
import java.util.concurrent.TimeUnit;

import static com.becomedataengineer.KafkaConfiguration.OUTPUT_TOPIC_NAME;

public class StatefulFlinkConsumer {

    public static void main(String[] args) throws Exception {
        Configuration conf = new Configuration();
        StreamExecutionEnvironment localEnvironment = StreamExecutionEnvironment.createLocalEnvironmentWithWebUI(conf);        localEnvironment.setParallelism(2);
        localEnvironment.enableCheckpointing(TimeUnit.SECONDS.toMillis(20));
        localEnvironment.setParallelism(1);

        KafkaSource<String> kafkaSource = KafkaSource.<String>builder()
                .setBootstrapServers("localhost:29092")
                .setStartingOffsets(OffsetsInitializer.earliest())
                .setValueOnlyDeserializer(new SimpleStringSchema())
                .setTopics(KafkaConfiguration.INPUT_TOPIC_NAME)
                .build();

        System.setProperty("user.timezone", "UTC");
        SingleOutputStreamOperator<String> kafkaDataStreamSource = localEnvironment.fromSource(kafkaSource,
                WatermarkStrategy.<String>forMonotonousTimestamps().withTimestampAssigner(
                        new VisitTimestampAssigner.Supplier()
                ),
                "Kafka Source").uid("kafka source");

        SingleOutputStreamOperator<Visit> visits = kafkaDataStreamSource.map((String inputRecord) -> Json.MAPPER.readValue(inputRecord, Visit.class));

        WindowedStream<Visit, String, TimeWindow> tumblingVisitWindow = visits.keyBy(Visit::getBrowser)
            .window(
                    TumblingEventTimeWindows.of(Time.minutes(5))
            ).allowedLateness(
                    Time.seconds(0)
            ).trigger(
                    EventTimeTrigger.create()
            );

        SingleOutputStreamOperator<String> windowOutput = tumblingVisitWindow.process(new ProcessWindowFunction<Visit, String, String, TimeWindow>() {
            @Override
            public void process(String browser,
                                ProcessWindowFunction<Visit, String, String, TimeWindow>.Context context,
                                Iterable<Visit> elements,
                                Collector<String> outputCollector) {
                try {
                    WindowOutput output = WindowOutput.valueOf(
                            context.window(), browser, IterableUtils.toStream(elements).count()
                    );
                    outputCollector.collect(Json.MAPPER.writeValueAsString(output));
                } catch (JsonProcessingException e) {
                    throw new RuntimeException(e);
                }
            }
        }).uid("windowOutput");

        KafkaSink<String> sink = KafkaSink.<String>builder()
                .setBootstrapServers("localhost:29092")
                .setRecordSerializer(KafkaRecordSerializationSchema.builder()
                        .setTopic(OUTPUT_TOPIC_NAME)
                        .setValueSerializationSchema(new SimpleStringSchema())
                        .build()
                )
                .setDeliveryGuarantee(DeliveryGuarantee.AT_LEAST_ONCE)
                .build();

        windowOutput.sinkTo(sink);
        localEnvironment.execute();
    }

}

File: 04-stateful-stream-processing/solution-proposal/python/e2/stateful_flink_consumer.py
#!/usr/bin/env python3
import datetime
import json
import os

from pyflink.common import SimpleStringSchema, WatermarkStrategy, Configuration, Time, Types
from pyflink.datastream import StreamExecutionEnvironment, RuntimeExecutionMode, WindowedStream, DataStream
from pyflink.datastream.connectors.kafka import KafkaSource, KafkaOffsetsInitializer, KafkaSink, \
    KafkaRecordSerializationSchema
from pyflink.datastream.window import TumblingEventTimeWindows, CountTrigger

from config import get_input_topic_name, get_output_topic_name
from partial_event_time_window_trigger import PartialEventTimeWindowTrigger
from visit import Visit
from visit_timestamp_assigner import VisitTimestampAssigner
from visit_window_processor import VisitWindowProcessor

# This configuration is very important
# Without it, you may encounter class loading-related errors, such as:
# Caused by: java.lang.ClassCastException: cannot assign instance of
#   org.apache.kafka.clients.consumer.OffsetResetStrategy to field
#   org.apache.flink.connector.kafka.source.enumerator.initializer.ReaderHandledOffsetsInitializer.offsetResetStrategy
#   of type org.apache.kafka.clients.consumer.OffsetResetStrategy
#   in instance of org.apache.flink.connector.kafka.source.enumerator.initializer.ReaderHandledOffsetsInitializer
config = Configuration()
config.set_string("classloader.resolve-order", "parent-first")
env = StreamExecutionEnvironment.get_execution_environment(configuration=config)
env.set_runtime_mode(RuntimeExecutionMode.STREAMING)
env.set_parallelism(2)
env.configure(config)

checkpoint_interval_30_seconds = 30000
env.enable_checkpointing(checkpoint_interval_30_seconds)
# Unlike PySpark, you have to define the JARs explicitly for Flink
env.add_jars(
    f"file://{os.getcwd()}/kafka-clients-3.2.3.jar",
    f"file://{os.getcwd()}/flink-connector-base-1.17.0.jar",
    f"file://{os.getcwd()}/flink-connector-kafka-1.17.0.jar"
)

kafka_source = KafkaSource.builder().set_bootstrap_servers('localhost:29092') \
    .set_group_id('bde_h06_e2') \
    .set_starting_offsets(KafkaOffsetsInitializer.earliest()) \
    .set_value_only_deserializer(SimpleStringSchema()).set_topics(get_input_topic_name()).build()

watermark_strategy = WatermarkStrategy.for_monotonous_timestamps().with_timestamp_assigner(VisitTimestampAssigner())
data_source = env.from_source(
    source=kafka_source,
    watermark_strategy=watermark_strategy,
    source_name="Kafka Source"
).uid("Kafka Source")


def map_json_to_visit(json_payload: str) -> Visit:
    event = json.loads(json_payload)
    event_time = int(datetime.datetime.fromisoformat(event['event_time']).timestamp())
    return Visit(visit_id=event['visit_id'], browser=event['browser'], event_time=event_time)


records_per_visit_counts: WindowedStream = data_source.map(map_json_to_visit) \
    .key_by(lambda visit: visit.browser).window(TumblingEventTimeWindows.of(Time.minutes(5))) \
    .allowed_lateness(0) \
    .trigger(PartialEventTimeWindowTrigger())

# must add the Types.STRING() to avoid serialization issues
window_output: DataStream = records_per_visit_counts.process(VisitWindowProcessor(), Types.STRING()).uid(
    "window output")

kafka_sink = KafkaSink.builder().set_bootstrap_servers("localhost:29092") \
    .set_record_serializer(KafkaRecordSerializationSchema.builder()
                           .set_topic(get_output_topic_name())\
                           .set_value_serialization_schema(SimpleStringSchema()).build()).build()

window_output.sink_to(kafka_sink)

env.execute('BDE irregular checkpoint example')
